Rework MersenneTwisterEngine using Ilya Yaroshenko's faster algorithm

WebDrake opened the pull request:
These patches update `MersenneTwisterEngine`'s implementation to use the new and much faster implementation @9il prepared for `mir.random`.  The standard 64-bit Mersenne Twister implementation is added as a side benefit.

The original Mersenne Twister algorithm (as per the reference C code by Matsumoto and Nishimura) first regenerates the contents of the entire state array, before looping over the resulting values and tempering them in order to generate the actual variates.  Once the entire state array has been iterated over, the array's values are then regenerated, and so on and so forth.

Ilya's implementation reworks this idea to intertwine update of a single entry in the state array with the tempering of another value to produce the next variate.  This ensures that all the registers concerned stay 'hot' in the CPU's memory and hence ensures significant speedup.

(Just as a mention: as an experiment while adapting this code for phobos I tried splitting apart the lines responsible for updating the internal state array from those responsible for tempering the next variate, so that they could be run sequentially instead of mixed up together; this resulted in a fairly significant speed hit.)

Ilya's code also introduces an extra tempering variable `d`, used in more recent Mersenne Twister implementations (Boost.Random, hap.random) but previously missing from Phobos' implementation.  This is needed in order to implement the standard 64-bit version of the Mersenne Twister, which is added in the second patch of this PR.

The first patch reworks Ilya's code to work as a drop-in replacement for the existing Phobos implementation, which is currently implemented as a forward range (although it should ideally be an input range, but that design flaw should be fixed as a separate issue).  There appears to be no significant speed hit from implementing the algorithm as a range rather than as a functor.  However, other aspects of the original design have required some rather intrusive changes to Ilya's code in order to get the full speed benefits without introducing breaking change.

The original MersenneTwisterEngine allows for implicit instantiation of generator instances without providing a seed.  This is handled by checking in the `front` and `popFront` method whether the generator has been properly initialized, and seeding it with the default seed if not.  However, these runtime checks on every single call result in a massive speed hit.  The current implementation therefore takes a number of steps in order to ensure that the internal state of the generator can be set
to its default-seeded values at compile time:

* all the internal state variables have been wrapped up in a nested `State` struct to facilitate generation;

* the internals of the `seed` and `popFront` methods have been separated out into CTFE'able private static methods (`seedImpl` and `popFrontImpl`) which take a reference to a `State` instance as input;

* a CTFE'able private static `defaultState` method has been added which generates a `State` instance matching that generated by instantiating the generator with the `defaultSeed`.

The `defaultState` method is then used to initialize the default value of the internal `State` instance at compile-time, replicating the effect of the original runtime seeding check.

These latter workarounds could be removed in future if the generator were updated to `@disable this();` and therefore always require explicit seeding, but this would be a breaking change and so is avoided for now.

WebDrake commented:
NOTE: the CTFE aspects of the current implementation have some known issues, which AFAICT are dmd's issues rather than those of the implementation per se.  I'm submitting for review anyway in the hope that it will help identify a solution or an alternative way of achieving the desired result.

In particular, the default initialization of the internal `state` field by `defaultState` works with ldc (checked with v1.1.0-beta6) but appears not to work with dmd (checked with both v2.072.2 and v2.071.2 [the latter matching the frontend version of ldc v1.1.0]): while the `state.z` value winds up set correctly, all the other values (including those in the `data` array) wind up being zero.

I've introduced a workaround patch to demonstrate the fundamental correctness of the new `MersenneTwisterEngine` implementation (it explicitly seeds some generator instances which rely on the "implicit" default seeding).  This 3rd patch should be removed before the code is considered merge-ready as otherwise the code cannot be guaranteed to work as a drop-in replacement.

WebDrake commented on std/random.d:
```diff
@@ -564,62 +572,135 @@ Parameters for the generator.
+       Constructs a MersenneTwisterEngine object by
+       duplicating the state of an existing instance
+    */
+    this(ref typeof(this) that)
```
A small note on the motivation for this constructor overload (which I should possibly add to the patch description if we decide to keep it).  A long-term goal is to remove the `.save` property of Phobos' random number generators: however, it is still useful to be able to deliberately deep-copy the state of an RNG in order to create a second generator that will produce identical results.

Having a constructor that copies an existing generator provides a way to "wrap up" the details of duplicating a generator in a way that can both be used directly (to construct a new instance) and also wrapped by other methods that might require it (`save` while it's still needed to avoid breaking change; `dup` has been proposed in future as a method that provides duplication without creating a forward range).

However, right now, this is creating a new feature that we might not want to support.  So it might be better to either make this constructor instance `private` or to remove it for now and just rework the internals of `save`.

WebDrake commented:
Re-pushed with fixes for the (unnecessary) nested import that caused circleci tests to fail, and a couple of other small issues that I noticed while editing the code:

```diff
diff --git a/std/random.d b/std/random.d
index 7fab38f..42c6389 100644
--- a/std/random.d
+++ b/std/random.d
@@ -534,8 +534,6 @@ struct MersenneTwisterEngine(UIntType, size_t w, size_t n, size_t m, size_t r,
                              uint l)
     if (isUnsigned!UIntType)
 {
-    import std.range.primitives;
-
     static assert(0 < w && w <= UIntType.sizeof * 8);
     static assert(1 <= m && m <= n);
     static assert(0 <= r && 0 <= u && 0 <= s && 0 <= t && 0 <= l);
@@ -580,7 +578,6 @@ Parameters for the generator.
         private UIntType z = 0;
         private UIntType index = void;
         private UIntType[n] data = void;
-
     }
 
     /// State variables used by the generator;
@@ -629,7 +626,7 @@ Parameters for the generator.
     */
     private static State defaultState() @safe pure nothrow @nogc
     {
-        //if (!__ctfe) assert(false);
+        if (!__ctfe) assert(false);
         State mtState;
         seedImpl(defaultSeed, mtState);
         return mtState;
```

WebDrake commented:
I don't understand the current circleci test failures -- they don't look like something to do with the PR ... ?  Can someone advise?

wilzbach commented:
> I don't understand the current circleci test failures -- they don't look like something to do with the PR ... ? Can someone advise?

Yep,  it's unrelated - the problem is in this line:

> curl: (22) The requested URL returned error: 403

(which is caused by sending more than 50 unauthenticated requests to the GitHub API within one hour) 
Subsequently `base_branch` isn't set and the `dmd` / `druntime` installation fails. 
IIRC there was a fallback to `master` in case of overusing the API, which seems not to work for 403s.
(the request to GitHub is needed to check against which branch a PR was opened against)

WebDrake commented:
Anything I can do to re-trigger the tests, or will they re-trigger automatically at some point?

9il commented:
@WebDrake Could you please add credits and link to Mir Random https://github.com/libmir/mir-random.

wilzbach commented:
> Anything I can do to re-trigger the tests, or will they re-trigger automatically at some point?

Unfortunately CircleCi won't restart itself without external action (the usual git push/sync events and PR close). AFAIK only maintainer can re-trigger a build manually, but it seems someone already did.
In any case I submitted https://github.com/dlang/phobos/pull/5012, which should fix this problem by using a "soft fallback".

joseph-wakeling-sociomantic commented on std/random.d:
```diff
@@ -630,93 +708,78 @@ Parameters for the generator.
+        if (conj < 0)
+            conj = index - m + n;
+        static if (d == UIntType.max)
+            z ^= (z >> u);
```
Interesting that code coverage analysis suggests this path is never taken; I would have expected it to be normal for the 32-bit `Mt19937`.

WebDrake replied:
Very odd that code coverage is suggesting this line is unused, as it should be for `UIntType == uint` i.e. for `Mt19937`.

WebDrake commented on std/random.d:
```diff
@@ -564,62 +570,134 @@ Parameters for the generator.
         }
-        popFront();
+        static if (is(UIntType == uint))
+            enum UIntType f = 1812433253;
```
It's odd that this branch is marked as not covered by codecov.io as it should be used for `Mt19937`.

9il replied:
this is compile time `enum`. It should not be covered if i am not wrong

WebDrake replied:
Ah, of course.  There's still another `static if` on L758-759 that codecov.io doesn't seem to like, though, which is weird.

WebDrake replied:
I ran the code coverage locally while running phobos unittests:

```
        |    private static void seedImpl(UIntType value, ref State mtState)
        |    {
        |        static if (w == UIntType.sizeof * 8)
        |        {
       8|            mtState.data[$-1] = value;
        |        }
        |        else
        |        {
        |            static assert(max + 1 > 0);
00000000|            mtState.data[$-1] = value % (max + 1);
        |        }
        |        static if (is(UIntType == uint))
00000000|            enum UIntType f = 1812433253;
        |        else static if (is(UIntType == ulong))
       8|            enum UIntType f = 6364136223846793005;
        |        else
        |            static assert(0, UIntType.stringof ~ " is not supported by MersenneTwisterEngine.");
    7488|        foreach_reverse (size_t i, ref e; mtState.data[0 .. $-1])
    2488|            e = f * (mtState.data[i + 1] ^ (mtState.data[i + 1] >> (w - 2))) + cast(UIntType)(n - (i + 1));
       8|        mtState.index = n-1;
        |
        |        // double popFront() to guarantee both
        |        // `_z` and `_y` are derived from the
        |        // newly set values in `data`
       8|        MersenneTwisterEngine.popFrontImpl(mtState);
       8|        MersenneTwisterEngine.popFrontImpl(mtState);
        |    }
```

The first `00000000` line makes sense, but the second doesn't considering that the `else` branch is marked as covered 8 times (and is also a compile-time `enum`).

9il replied:
Very strange. Maybe it is always 64 bit?

WebDrake replied:
I'll check, but I don't think there is any sane way that can be true; `Mt19937` (uses `uint`) and `Mt19937_64` (uses `ulong`) are both instantiated and used.

9il commented on std/random.d:
```diff
@@ -527,20 +527,25 @@ alias MinstdRand = LinearCongruentialEngine!(uint, 48271, 0, 2147483647);
+                             uint u, UIntType d,
+                             uint s, UIntType b,
+                             uint t, UIntType c,
+                             uint l)
```
Shift parameters are ok to be `uint`. No strong reasons.

WebDrake replied:
OK, cool, thanks!  I'm trying to find ways in which to minimize the diff so as to make it most informative, but TBH I think having fixed-size template parameters is probably preferable.

WebDrake replied:
Hmm, on checking, the C++11 standard explicitly defines these template parameters as `size_t`. While it may not make any practical difference, I don't see any reason to diverge from that spec, and it minimizes the different with phobos original, so I'm going to restore the `size_t`.

WebDrake commented:
I've pushed a bunch of small fixes and extra tests, including the credit to Ilya and mir.random.  Most of these will be squashed into the earlier patches; a couple are extra tests that should be placed before the new-algorithm patches in the final patchset.

WebDrake commented:
I see circleci is up to another of its evil tricks ... anyway, feedback welcome on the current code.

WebDrake commented on std/random.d:
```diff
@@ -806,12 +941,17 @@ alias Mt19937 = MersenneTwisterEngine!(uint, 32, 624, 397, 31,
     foreach (R; std.meta.AliasSeq!(MT!(uint, 32), MT!(ulong, 32), MT!(ulong, 48), MT!(ulong, 64)))
+    {
         auto a = R();
+        a.seed(a.defaultSeed); // checks that some alternative paths in `seed` are utilized
```
A note on this change: ideally it would be nice to include checks on `a.front` (before & after `popFrontN(9999)`).  I don't have any values from a reference implementation to compare to right now, but I'll try to follow up on this soon-ish.

WebDrake commented:
@9il my comparison is made based on g++ 5.4.0.  More specifically, compare the output of the following D code (slightly adapted from one of the phobos unittests):

```D
unittest
{
    alias MT(UIntType, uint w) = MersenneTwisterEngine!(UIntType, w, 624, 397, 31,
                                                        0x9908b0df, 11, 0xffffffff, 7,
                                                        0x9d2c5680, 15,
                                                        0xefc60000, 18, 1812433253);

    foreach (R; std.meta.AliasSeq!(MT!(uint, 32), MT!(ulong, 32), MT!(ulong, 48), MT!(ulong, 64)))
    {
        static if (R.wordSize == 48) static assert(R.max == 0xFFFFFFFFFFFF);
        auto a = R();
        a.seed(a.defaultSeed); // checks that some alternative paths in `seed` are utilized
        import std.stdio : writeln;
        writeln(a.front);  // for mir.random replace these 3 lines with
        a.popFront();      // writeln(a());
        writeln(a.front);  // writeln(a());
        writeln();
    }
}
```

to this C++11 program:

```C++
#include <cinttypes>
#include <iostream>
#include <random>
#include <type_traits>

template<class UIntType, size_t w>
void mt_result ()
{
    std::mersenne_twister_engine<
        UIntType, w, 624, 397, 31,
        0x9908b0df, 11, 0xffffffff, 7,
        0x9d2c5680, 15,
        0xefc60000, 18, 1812433253> gen;

    gen.seed(std::mt19937::default_seed);
    std::cout << gen() << std::endl;
    std::cout << gen() << std::endl;
    std::cout << std::endl;
}

int main ()
{
    mt_result<uint32_t, 32>();
    mt_result<uint64_t, 32>();
    mt_result<uint64_t, 48>();
    mt_result<uint64_t, 64>();
}
```

I've submitted a PR to update the `mir.random` Mersenne Twister's template signature to match C++11 precisely.  From there it should be easier to make comparisons and pinpoint the source of the problem.

So far as I can work out from comparison to MT spec, the essentials of the problem are that the word size `w` is not fully taken into account when seeding the generator nor when returning tempered variates.

9il commented:
> I've submitted a PR to update the mir.random Mersenne Twister's template signature to match C++11 precisely. From there it should be easier to make comparisons and pinpoint the source of the problem.

Thanks! Merged

WebDrake commented:
Thanks to some fast fixes from @9il the word-size issue is now addressed, and a unittest has been added to validate comparison to expected behaviour when the word size occupies less than a full `UIntType`.

How would everyone like to proceed from here?  If I don't have more detailed feedback by tomorrow I think it may be better to squash everything down once more to a clean patchset.

WebDrake commented on std/random.d:
```diff
@@ -549,77 +554,151 @@ Parameters for the generator.
+    {
+        private UIntType y = void;
+        private UIntType z = 0;
+        private UIntType index = void;
```
This should be a `size_t`.

WebDrake commented on std/random.d:
```diff
@@ -527,20 +528,24 @@ alias MinstdRand = LinearCongruentialEngine!(uint, 48271, 0, 2147483647);
     static assert(0 <= r && 0 <= u && 0 <= s && 0 <= t && 0 <= l);
     static assert(r <= w && u <= w && s <= w && t <= w && l <= w);
     static assert(0 <= a && 0 <= b && 0 <= c);
+    static assert(n < UIntType.max);
```
Should probably be a comparison to `sizediff_t.max` to make sure the index values will always fall within acceptable bounds.

WebDrake commented:
OK, with the major issues dealt with I've squashed everything down to a clean set of patches that I think can be considered "final", bar feedback and bar the resolution of the CTFE issue related to the `WORKAROUND` patch.

I'll submit a separate issue report for the range-based seeding concern.

WebDrake commented:
Range-based seeding mechanism issue is now reported here:
https://issues.dlang.org/show_bug.cgi?id=17068

WebDrake commented:
> the resolution of the CTFE issue related to the `WORKAROUND` patch.

I've tracked it down: it's due to the `void` initialization of fields in the `State` struct clashing with the default initialization from `defaultState()`.  Updated PR arriving shortly.

WebDrake commented:
The updated patches remove the `void` default-initialization of fields in the `State` struct that was clashing with the results of the `defaultState()` CTFE function.  The unittest workaround patch has therefore been removed.

This patchset can now be considered without known issues & therefore 100% ready for review.

The void-initialization issue has been reported here: https://issues.dlang.org/show_bug.cgi?id=17073

WebDrake commented:
OK, one last change: I've stripped out any unnecessary documentation and function-declaration changes introduced in the main patch (including ordering of struct methods), and made one small documentation change (regarding seeding) that _is_ necessary.  I've also rebased on current master.

This should minimize the diff from the original to what is strictly necessary.

WebDrake commented:
Given no feedback in over a week, I've made another couple of small changes that I think improve the code and patchset:

* `front` has been made `const` (which is possible now that `front` no longer has any involvement in default-seeding the generator);

* I've added a note in the ddoc for the generator state that this implementation iterates _backwards_ rather than forwards through the state array, and similarly noted this in the commit message body.

The patches have also been rebased on master.  The diff to `std/random.d` is as follows:

```diff
diff --git a/std/random.d b/std/random.d
index f433c36..b60101b 100644
--- a/std/random.d
+++ b/std/random.d
@@ -578,7 +578,12 @@ Parameters for the generator.
     /// used by the generator
     private struct State
     {
-        /// State array of the generator
+        /// State array of the generator.  This
+        /// is iterated through backwards (from
+        /// last element to first), providing a
+        /// few extra compiler optimizations by
+        /// comparison to the forward iteration
+        /// used in most implementations.
         UIntType[n] data;
 
         /// Cached copy of most recently updated
@@ -781,7 +786,7 @@ Parameters for the generator.
 /**
    Returns the current random value.
  */
-    @property UIntType front() @safe pure nothrow @nogc
+    @property UIntType front() @safe const pure nothrow @nogc
     {
         return this.state.front;
     }
```

WebDrake commented:
@JackStouffer given that you seem to have taken responsibility for tagging this PR, is it possible to add the `Enhancement` tag (since it also introduces a new symbol, `Mt19937_64`) and the `BugFix` tag (since it corrects some behaviour of the Mersenne Twister implementation compared to what was previously in Phobos -- specifically, related to the word size `w`) ... ?

WebDrake commented:
> since it corrects some behaviour of the Mersenne Twister implementation compared to what was previously in Phobos -- specifically, related to the word size `w`

I realized this was not noted in the main commit message, so I've added a note on this too.  Genuinely the last change I can think of (absent any review feedback).

@JackStouffer thanks for adding the tags :-)

JackStouffer commented:
Pinging people who are qualified to review this based on `git blame` @braddr @jpf91 @9il

WebDrake commented:
Folks, I'm having a bit of a difficult time understanding why these tests are failing.  It looks like they are either due to integration issues with the CI or to problems elsewhere in Phobos ... ?

In any case, ping regarding review of this PR?

WebDrake commented:
Rebased on master.  Hopefully should get a test pass now that the CI issues have been dealt with.

WebDrake commented:
Again, the failure doesn't look like anything to do with this patchset. :-\

andralex commented on std/random.d:
```diff
@@ -549,21 +551,67 @@ Parameters for the generator.
 
+    /// Bitmasks used in the 'twist' part of the algorithm
+    private enum UIntType lowerMask = (cast(UIntType) 1u << r) - 1;
+    private enum UIntType upperMask = (~lowerMask) & this.max; // ditto
```
perhaps use a comma instead of repeating `private enum UIntType`?

andralex commented on std/random.d:
```diff
@@ -549,21 +551,67 @@ Parameters for the generator.
+    private enum UIntType upperMask = (~lowerMask) & this.max; // ditto
+
+    /// Collection of all state variables
+    /// used by the generator
```
Please use block comments instead.

andralex commented on std/random.d:
```diff
@@ -549,21 +551,67 @@ Parameters for the generator.
+        /// last element to first), providing a
+        /// few extra compiler optimizations by
+        /// comparison to the forward iteration
+        /// used in most implementations.
```
here, there, and everywhere

andralex commented on std/random.d:
```diff
@@ -572,36 +620,60 @@ Parameters for the generator.
+        if (!__ctfe) assert(false);
+        State mtState;
+        seedImpl(defaultSeed, mtState);
+        return mtState;
```
eh, we need -cov to cover CTFE...

andralex commented on std/random.d:
```diff
@@ -572,36 +620,60 @@ Parameters for the generator.
+    */
+    private static void seedImpl(UIntType value, ref State mtState)
+    {
+        mtState.data[$-1] = value;
```
space around operators

andralex replied:
cc @wilzbach

andralex commented on std/random.d:
```diff
@@ -572,36 +620,60 @@ Parameters for the generator.
         {
-            static assert(max + 1 > 0);
-            mt[0] = value % (max + 1);
+            mtState.data[$-1] &= this.max;
```
spaces

andralex commented on std/random.d:
```diff
@@ -572,36 +620,60 @@ Parameters for the generator.
         }
-        popFront();
+
+        mtState.index = n-1;
```
and more spaces throughout

andralex commented on std/random.d:
```diff
@@ -549,21 +551,67 @@ Parameters for the generator.
+    /// State variables used by the generator;
+    /// initialized to values equivalent to
+    /// explicitly seeding the generator with
+    /// `defaultSeed`
```
BTW these should not be doc comments, right?

WebDrake replied:
Everything commented with `///` was intended as a ddoc comment. I assumed it was better to provide the info and leave it to the doc generator to decide what symbols' doc to expose.  (IIRC as private symbols they should not end up in generated docs by default.)

Would you prefer these be regular comments?

andralex replied:
@WebDrake they may confuse the reader (was this symbol supposed to be exposed? if not why is this a ddoc comment? etc). At any rate, it may be nice to transform these into block comments. Then we can switch from ddoc to regular comment with a single character change.

WebDrake replied:
OK, makes sense. I'll follow up on this tomorrow.

JackStouffer commented:
Also, please add a changelog entry.

WebDrake commented:
Thanks for the detailed feedback, everyone :-) I'll try to follow up by tomorrow evening (today is a busy day).

WebDrake commented on std/random.d:
```diff
@@ -527,16 +528,17 @@ alias MinstdRand = LinearCongruentialEngine!(uint, 48271, 0, 2147483647);
+                             UIntType a, size_t u, UIntType d, size_t s,
                              UIntType b, size_t t,
-                             UIntType c, size_t l)
+                             UIntType c, size_t l, UIntType f)
```
Yes, technically it's a breaking change. I'd consider it quite unlikely that anyone is using the template directly, though (and if they are it is probably not safe to do so given the missing parameters).

I'll make sure this is noted prominently in the changelog, though.

andralex replied:
OK, I'm fine with this. cc @MartinNowak

WebDrake commented:
Rebased on master, made the requested tweaks, and added a changelog entry.

Ddoc comments for private symbols have been replaced by comments

```D
/*
   Like this
*/
```

so they can easily be re-ddoc'd in future if that's desirable.  Multiline comments that will never be ddoc have been replaced by comments

```D
/* Like
   this */
```

WebDrake commented:
Overall changes to `std.random`:

```diff
diff --git a/std/random.d b/std/random.d
index b60101b..f3b1b45 100644
--- a/std/random.d
+++ b/std/random.d
@@ -570,47 +570,59 @@ Parameters for the generator.
     /// The default seed value.
     enum UIntType defaultSeed = 5489u;
 
-    /// Bitmasks used in the 'twist' part of the algorithm
-    private enum UIntType lowerMask = (cast(UIntType) 1u << r) - 1;
-    private enum UIntType upperMask = (~lowerMask) & this.max; // ditto
+    // Bitmasks used in the 'twist' part of the algorithm
+    private enum UIntType lowerMask = (cast(UIntType) 1u << r) - 1,
+                          upperMask = (~lowerMask) & this.max;
 
-    /// Collection of all state variables
-    /// used by the generator
+    /*
+       Collection of all state variables
+       used by the generator
+    */
     private struct State
     {
-        /// State array of the generator.  This
-        /// is iterated through backwards (from
-        /// last element to first), providing a
-        /// few extra compiler optimizations by
-        /// comparison to the forward iteration
-        /// used in most implementations.
+        /*
+           State array of the generator.  This
+           is iterated through backwards (from
+           last element to first), providing a
+           few extra compiler optimizations by
+           comparison to the forward iteration
+           used in most implementations.
+        */
         UIntType[n] data;
 
-        /// Cached copy of most recently updated
-        /// element of `data` state array, ready
-        /// to be tempered to generate next
-        /// `front` value
+        /*
+           Cached copy of most recently updated
+           element of `data` state array, ready
+           to be tempered to generate next
+           `front` value
+        */
         UIntType z;
 
-        /// Most recently generated random variate
+        /*
+           Most recently generated random variate
+        */
         UIntType front;
 
-        /// Index of the entry in the `data`
-        /// state array that will be twisted
-        /// in the next `popFront()` call
+        /*
+           Index of the entry in the `data`
+           state array that will be twisted
+           in the next `popFront()` call
+        */
         size_t index;
     }
 
-    /// State variables used by the generator;
-    /// initialized to values equivalent to
-    /// explicitly seeding the generator with
-    /// `defaultSeed`
+    /*
+       State variables used by the generator;
+       initialized to values equivalent to
+       explicitly seeding the generator with
+       `defaultSeed`
+    */
     private State state = defaultState();
-    // NOTE: the above is a workaround to ensure
-    // backwards compatibility with the original
-    // implementation, which permitted implicit
-    // construction.  With `@disable this();`
-    // it would not be necessary.
+    /* NOTE: the above is a workaround to ensure
+       backwards compatibility with the original
+       implementation, which permitted implicit
+       construction.  With `@disable this();`
+       it would not be necessary. */
 
 /**
    Constructs a MersenneTwisterEngine object.
@@ -652,13 +664,13 @@ Parameters for the generator.
     */
     private static void seedImpl(UIntType value, ref State mtState)
     {
-        mtState.data[$-1] = value;
+        mtState.data[$ - 1] = value;
         static if (this.max != UIntType.max)
         {
-            mtState.data[$-1] &= this.max;
+            mtState.data[$ - 1] &= this.max;
         }
 
-        foreach_reverse (size_t i, ref e; mtState.data[0 .. $-1])
+        foreach_reverse (size_t i, ref e; mtState.data[0 .. $ - 1])
         {
             e = f * (mtState.data[i + 1] ^ (mtState.data[i + 1] >> (w - 2))) + cast(UIntType)(n - (i + 1));
             static if (this.max != UIntType.max)
@@ -667,11 +679,11 @@ Parameters for the generator.
             }
         }
 
-        mtState.index = n-1;
+        mtState.index = n - 1;
 
-        // double popFront() to guarantee both `mtState.z`
-        // and `mtState.front` are derived from the newly
-        // set values in `mtState.data`
+        /* double popFront() to guarantee both `mtState.z`
+           and `mtState.front` are derived from the newly
+           set values in `mtState.data` */
         MersenneTwisterEngine.popFrontImpl(mtState);
         MersenneTwisterEngine.popFrontImpl(mtState);
     }
@@ -714,9 +726,9 @@ Parameters for the generator.
             throw new Exception(s);
         }
 
-        // double popFront() to guarantee both `mtState.z`
-        // and `mtState.front` are derived from the newly
-        // set values in `mtState.data`
+        /* double popFront() to guarantee both `mtState.z`
+           and `mtState.front` are derived from the newly
+           set values in `mtState.data` */
         MersenneTwisterEngine.popFrontImpl(mtState);
         MersenneTwisterEngine.popFrontImpl(mtState);
     }
@@ -729,22 +741,24 @@ Parameters for the generator.
         this.popFrontImpl(this.state);
     }
 
-    /// Internal implementation of `popFront()`, which
-    /// can be used with an arbitrary `State` instance
+    /*
+       Internal implementation of `popFront()`, which
+       can be used with an arbitrary `State` instance
+    */
     private static void popFrontImpl(ref State mtState)
     {
-        // This function blends two nominally independent
-        // processes: (i) calculation of the next random
-        // variate `mtState.front` from the cached previous
-        // `data` entry `z`, and (ii) updating the value
-        // of `data[index]` and `mtState.z` and advancing
-        // the `index` value to the next in sequence.
-        //
-        // By interweaving the steps involved in these
-        // procedures, rather than performing each of
-        // them separately in sequence, the variables
-        // are kept 'hot' in CPU registers, allowing
-        // for significantly faster performance.
+        /* This function blends two nominally independent
+           processes: (i) calculation of the next random
+           variate `mtState.front` from the cached previous
+           `data` entry `z`, and (ii) updating the value
+           of `data[index]` and `mtState.z` and advancing
+           the `index` value to the next in sequence.
+
+           By interweaving the steps involved in these
+           procedures, rather than performing each of
+           them separately in sequence, the variables
+           are kept 'hot' in CPU registers, allowing
+           for significantly faster performance. */
         sizediff_t index = mtState.index;
         sizediff_t next = index - 1;
         if (next < 0)
@@ -776,10 +790,10 @@ Parameters for the generator.
         mtState.z = mtState.data[index] = e;
         mtState.index = next;
 
-        // technically we should take the lowest `w`
-        // bits here, but if the tempering bitmasks
-        // `b` and `c` are set correctly, this should
-        // be unnecessary
+        /* technically we should take the lowest `w`
+           bits here, but if the tempering bitmasks
+           `b` and `c` are set correctly, this should
+           be unnecessary */
         mtState.front = z;
     }
 
```

WebDrake commented:
Updated the changelog entry slightly to stress the breaking change w.r.t. the `MersenneTwisterEngine` template.

JackStouffer commented:
Auto-merge toggled on

WebDrake commented:
Thanks everyone! :-)

I realized that the changelog was missing one important note, which I've added in a follow-up PR: https://github.com/dlang/phobos/pull/5141
